source/agora/consensus/validation/Block.d

```diff
@@ -113,6 +119,282 @@ public string isInvalidReason (const ref Block block, in ulong prev_height,
     return null;
 }
 
+/*******************************************************************************
+
+    Check the validity of a genesis block
+
+    Follow the Block rules except for the following
+        - Block height is 0
+        - The previous block hash of the block must be empty
+        - The number of transactions in the block is above zero and less than
+          or equal to Block.TxsInBlock
+        - All the transactions pass validation, which implies:
+            - Transaction type in block should only be Payment or Freeze
+            - The inputs of the transaction(s) are empty
+            - There is one or more outputs of the transaction
+        - All the enrollments pass validation, which implies:
+            - The enrollments refer to freeze tx's in this block
+
+    Params:
+        block = The genesis block to check
+
+    Returns:
+        `null` if the genesis block is valid, otherwise a string explaining
+        the reason it is invalid.
+
+*******************************************************************************/
+
+public string isValidGenesisBlock (immutable Block* block) nothrow @safe
```

Comment:
This should take `const ref`, not `immutable *`

Reply:
Then you wouldn't need the `serializeFull.deserialize` workaround in `testIsValidGenesisBlock`.

Reply:
Alternatively just take it by value if it's only used in one place.

Reply:
https://github.com/bpfkorea/agora/blob/72542f5cc9e0c624176f3b809670542ce4abbc87/source/agora/consensus/Genesis.d#L23-L26
Used as declared.
If I want to cast as `immutable` to `const` shouldn't I have to do `GenesisBlock.serializeFull.deserializeFull!Block`?

Reply:
You don't have to cast anything. Just assign the genesis block, and then pass `*gen_block` to the validation function